distribution/lib/Standard/Test/0.0.0-dev/src/Bench.enso

```diff
@@ -1,7 +1,71 @@
 from Standard.Base import all
 import Standard.Base.Runtime.Ref.Ref
 
+
+type Bench_Options
+    ## PRIVATE
+    Impl iter_size num_iters
+
+    size : Integer -> Bench_Options
+    size self v = Bench_Options.Impl v self.num_iters
+
+    iter : Integer -> Bench_Options
+    iter self v = Bench_Options.Impl self.iter_size v
+
+    to_text self = "[iter_size=" + self.iter_size.to_text + ", num_iters=" + self.num_iters.to_text + "]"
+
+type Bench_Builder
+    ## PRIVATE
+    Impl builder
+
+    group : Text -> Bench_Options -> (Group_Builder -> Any) -> Any
+    group self (name:Text) (configuration:Bench_Options) fn =
+        b = Vector.new_builder
+        fn (Group_Builder.Impl b)
+        self.builder.append <| Bench.Group name configuration b.to_vector
+
+type Group_Builder
+    ## PRIVATE
+    Impl builder
+
+    specify : Text -> Any -> Bench
+    specify self (name:Text) ~benchmark =
+        self.builder.append <| Bench.Spec name (_ -> benchmark)
+
+
 type Bench
+    All (groups : Vector Bench)
+    Group (name:Text) (configuration:Bench_Options) (specs : Vector Bench)
+    Spec (name:Text) (code : Any -> Any)
+
+    build : (Bench_Builder -> Any) -> Bench
+    build fn =
```

Comment:
I was experimenting with various ways to construct the benchmark cases, but I have found nothing better than old good builder pattern. I tried various tricks with `State`, but at the end `Vector.new_builder` (with its mutability) seems to work the best and without major surprises.

```haskell
from Standard.Base import all
from Standard.Test import Bench

all = Bench.build builder->
    builder.group "Group1" (Bench.options.iter 3 . size 1) group->
        group.specify "bench1"
            1 * 1
        group.specify "bench2"
            1 * 2
    builder.group "group2" (Bench.options.iter 5 . size 1) group->
        group.specify "another1"
            2 * 1
        group.specify "another2"
            2 * 2

main = all.run_main
```
if this is find acceptable, then we should change our `Test_Suite` harness to use the same pattern. For integration with VSCode it is important to collect tests first (even without executing them), present them to the user and allow user to choose one of them to be executed.

Reply:
I was thinking about a similar pattern. This suggestion looks good. The only thing that remains to be solved is how to implement `Bench.run_main` such that when run from CLI, it will behave in the old way, and when run from JMH, it will be a no-op. I believe that this can be tweaked in the JMH. Generally, we just need to think about how to achieve that once this script is invoked from CLI, it will immediately run all the benchmarks and when invoked from JMH, it will gather the descriptions.

TL;DR; I believe the builder pattern, similar to the one we are already using for our Enso tests, is the only way to go.

Reply:
>  when run from CLI, it will behave in the old way

Why it should behave the old way? Is there any benefit in _not collecting_ all the cases first? I can only think of more memory needed to hold all the cases initialized, but even that can be solved by `@Before` (or `@Setup`) and `@After` annotations.

I certainly need the tests to behave in such a way - _collect all the cases first_ and only then let the VSCode invoke (for example) one of them.

Reply:
I really like collecting the cases, as you say - we can then easily do a 'dry run' and allow more interactive test selection. Moreover, knowing the amount of tests beforehand gives us first steps towards adding some progress indicators (% of tests already completed, maybe ETA) - I love progress bars so I'd love to see this (ofc this is not a very important feature but a cool nice to have one).

For the test suite there's one thing to consider - our tests tend to do some 'preparation' before running the particular test cases (`Test.specify` branches) - some of it inside of `Test.group` and even some outside of it. That preparation tends to be side-effecting (e.g. 
 establishing DB connection, creating some base tables in a DB) and takes some time (creating a table over a connection tends to be in the magnitude of 300ms, not 3ms IIRC).

When gathering just descriptions - we probably don't want to run this preparatory stuff as it would only be a waste of time. But we cannot easily 'fold' it into `Test.specify` because that would force the preparation to be re-run for every single test case - I'm not sure what will be the cost of that - we may have to measure it.

I think ideally we'd need to refactor the tests in such a way that the structure can be gathered without running these preparatory actions. It may require some work to achieve that.

Reply:
Running the _"preparation"_ only when needed is an essential part of testing API. That's why I'd like the libraries team to design it, at the end. I can think of the _"JUnit style"_ - e.g. adding 
```ruby
type Group_Builder
    before : Function -> ???
    after: Function -> ???
```
each group would call all registered `before` handlers before running `specify` and then it would call `after`. However how to access the results of `before` in `specify` (in a side-effect free language) remains a mystery to me!?

However we may not need any special construct. Enough to specify lazy value. E.g. define
```ruby
type My_Data
    Connection ~sql
```
and then create an instance in `group` scope and access the `instance.sql` only when needed. It will initialize just ones, at first warmup iteration.

How to handle cleanup isn't completely clear to me. Could `Managed_Resource` help somehow? Let's hope it would and let's thus leave the topic of _before&after_ for later. It is not essential to running of a _single benchmark_ that this PR tackles.

Reply:
I imagine we could have `specify` take an additional argument which will be filled in by `before`?

i.e.

```
builder.before <|
    h1 = ...
    h2 = ...
    My_Helpers.Instance h1 h2

builder.specify "foo" helpers->
    helpers.h1.foo_bar helpers.h2
```
what do you think?
It's not the most pretty IMHO but it could work ok.

Reply:
Yet another suggestion is to have `before` as part of `Bench_Options` and specify it as function to function.

```diff
diff --git a/distribution/lib/Standard/Test/0.0.0-dev/src/Bench.enso b/distribution/lib/Standard/Test/0.0.0-dev/src/Bench.enso
--- a/distribution/lib/Standard/Test/0.0.0-dev/src/Bench.enso	(revision 74ac8d759e54c9880477f06eb8907d83b8107580)
+++ b/distribution/lib/Standard/Test/0.0.0-dev/src/Bench.enso	(date 1689700961515)
@@ -4,13 +4,16 @@
 
 type Bench_Options
     ## PRIVATE
-    Impl iter_size num_iters
+    Impl iter_size num_iters before_func
 
     size : Integer -> Bench_Options
-    size self v = Bench_Options.Impl v self.num_iters
+    size self v = Bench_Options.Impl v self.num_iters self.before_func
 
     iter : Integer -> Bench_Options
-    iter self v = Bench_Options.Impl self.iter_size v
+    iter self v = Bench_Options.Impl self.iter_size v self.before_func
+
+    before : (Nothing -> Any) -> Bench_Options
+    before self fn = Bench_Options.Impl self.iter_size self.num_iters fn
 
     to_text self = "[iter_size=" + self.iter_size.to_text + ", num_iters=" + self.num_iters.to_text + "]"
 
@@ -45,7 +48,7 @@
         Bench.All b.to_vector
 
     options : Bench_Options
-    options = Bench_Options.Impl -1 -1
+    options = Bench_Options.Impl -1 -1 Nothing
 
     fold : Any -> (Any -> Bench -> Bench -> Any) -> Any
     fold self value fn = case self of
@@ -60,7 +63,9 @@
         self.fold Nothing _-> g-> s->
             c = g.configuration
             IO.println <| "Benchmarking  " + s.name + " configuration: " + c.to_text
-            Bench.measure (s.code 0) s.name c.iter_size c.num_iters
+            data = c.before_func Nothing
+            IO.println <| "Before data = " + data.to_text
+            Bench.measure (s.code data) s.name c.iter_size c.num_iters
             IO.println <| "Benchmarking of " + s.name + " finished"
 
     ## Measure the amount of time it takes to execute a given computation.

```

And usage:
```
from Standard.Base import all
import Standard.Test.Bench.Bench
import Standard.Test.Bench.Bench_Options

before_func =
    # Do some side effects?
    # ...
    IO.println "[mylog] In before_func"
    [1,2,3,4,5]

tmp = Bench.options . size 10
options = Bench.options . size 10 . iter 5 . before (_ -> before_func)

all = Bench.build builder->
    builder.group "group_1" options group_builder->
        group_builder.specify "bench_1" <|
            vec -> vec + [10]

main =
    all.run_main

```

Reply:
Yet another suggestion is to have `before` as part of `Bench_Options` and specify it as function to function.

```diff
diff --git a/distribution/lib/Standard/Test/0.0.0-dev/src/Bench.enso b/distribution/lib/Standard/Test/0.0.0-dev/src/Bench.enso
--- a/distribution/lib/Standard/Test/0.0.0-dev/src/Bench.enso	(revision 74ac8d759e54c9880477f06eb8907d83b8107580)
+++ b/distribution/lib/Standard/Test/0.0.0-dev/src/Bench.enso	(date 1689700961515)
@@ -4,13 +4,16 @@
 
 type Bench_Options
     ## PRIVATE
-    Impl iter_size num_iters
+    Impl iter_size num_iters before_func
 
     size : Integer -> Bench_Options
-    size self v = Bench_Options.Impl v self.num_iters
+    size self v = Bench_Options.Impl v self.num_iters self.before_func
 
     iter : Integer -> Bench_Options
-    iter self v = Bench_Options.Impl self.iter_size v
+    iter self v = Bench_Options.Impl self.iter_size v self.before_func
+
+    before : (Nothing -> Any) -> Bench_Options
+    before self fn = Bench_Options.Impl self.iter_size self.num_iters fn
 
     to_text self = "[iter_size=" + self.iter_size.to_text + ", num_iters=" + self.num_iters.to_text + "]"
 
@@ -45,7 +48,7 @@
         Bench.All b.to_vector
 
     options : Bench_Options
-    options = Bench_Options.Impl -1 -1
+    options = Bench_Options.Impl -1 -1 Nothing
 
     fold : Any -> (Any -> Bench -> Bench -> Any) -> Any
     fold self value fn = case self of
@@ -60,7 +63,9 @@
         self.fold Nothing _-> g-> s->
             c = g.configuration
             IO.println <| "Benchmarking  " + s.name + " configuration: " + c.to_text
-            Bench.measure (s.code 0) s.name c.iter_size c.num_iters
+            data = c.before_func Nothing
+            IO.println <| "Before data = " + data.to_text
+            Bench.measure (s.code data) s.name c.iter_size c.num_iters
             IO.println <| "Benchmarking of " + s.name + " finished"
 
     ## Measure the amount of time it takes to execute a given computation.

```

And usage:
```
from Standard.Base import all
import Standard.Test.Bench.Bench
import Standard.Test.Bench.Bench_Options

before_func =
    # Do some side effects?
    # ...
    IO.println "[mylog] In before_func"
    [1,2,3,4,5]

tmp = Bench.options . size 10
options = Bench.options . size 10 . iter 5 . before (_ -> before_func)

all = Bench.build builder->
    builder.group "group_1" options group_builder->
        group_builder.specify "bench_1" <|
            vec -> vec + [10]

main =
    all.run_main

```

Reply:
Discussing `before` is completely fine, but please leave the implementation for another PR. Let's focus on getting this PR merged during this iteration with the goal to have an `sbt` command to invoke (some of) the Enso benchmarks and get `bench-result.xml` compatible output.